aten/src/ATen/core/boxing/KernelFunction.h

```diff
@@ -0,0 +1,397 @@
+    TORCH_INTERNAL_ASSERT(func != nullptr, "Kernel function cannot be nullptr");
+
+    return makeFromUnboxedFunctor<AllowLegacyTypes>(
+      std::make_shared<detail::WrapRuntimeKernelFunctor<guts::decay_t<FuncType>>>(func)
+    );
+  }
+
+  /**
+   * Create a KernelFunction from an unboxed lambda.
+   *
+   * Example:
+   *
+   * > KernelFunction func = KernelFunction::makeFromUnboxedLambda(
+   * >      [] (Tensor a, bool b) -> Tensor {...});
+   */
+  template<bool AllowLegacyTypes = false, class Lambda>
+  static KernelFunction makeFromUnboxedLambda(Lambda&& lambda) {
+    static_assert(guts::is_functor<guts::decay_t<Lambda>>::value, "Tried to call KernelFunction::makeFromUnboxedLambda with a non-lambda type.");
+
+    return makeFromUnboxedFunctor<AllowLegacyTypes>(
+      std::make_shared<detail::WrapRuntimeKernelFunctor<guts::decay_t<Lambda>>>(std::forward<Lambda>(lambda))
+    );
+  }
+
+private:
+
+  explicit KernelFunction(std::function<std::shared_ptr<OperatorKernel>()> functorFactory, std::shared_ptr<OperatorKernel> functor, BoxedKernelFunction* boxed_kernel_func, void* unboxed_kernel_func)
+  : functorFactory_(std::move(functorFactory))
+  , functor_(std::move(functor))
+  , boxed_kernel_func_(boxed_kernel_func)
+  , unboxed_kernel_func_(unboxed_kernel_func)
+  {}
+
+  OperatorKernel* getFunctor_() const {
+    if (functor_.get() == nullptr) {
+      if (!functorFactory_) {
+        return nullptr;
+      }
+      functor_ = functorFactory_();
+    }
+    return functor_.get();
+  }
+
+  // If the operator has an unboxed_kernel_func, then either
+  // functorFactory_ or functor_ must be set, possibly both.
+  // If functor_ is not set but functorFactory_ is, we will create
+  // functor_ by calling functorFactory_ the first time it is needed.
+  // We use this indirection because many KernelFunctions are created
+  // at static initialization time but are created with functors that
+  // store Tensor and we can't call the Tensor() constructor at static
+  // initialization time yet (SIOF). So these register with a
+  // functorFactory_ instead of a functor_ and will be initialized
+  // on the first call to the KernelFunction.
+  std::function<std::shared_ptr<OperatorKernel>()> functorFactory_;
+  mutable std::shared_ptr<OperatorKernel> functor_;
+
+  BoxedKernelFunction* boxed_kernel_func_;
+  void* unboxed_kernel_func_;
+};
+
+namespace detail {
+template<class Return, class... Args>
+struct boxAndCallBoxedFunc final {
+  static Return call(KernelFunction::BoxedKernelFunction* boxed_kernel_func, OperatorKernel* functor, Args... args) {
+    // TODO Reuse stack vector instead of allocating?
+    std::vector<IValue> stack {std::forward<Args>(args)...};
+
+    (*boxed_kernel_func)(functor, &stack);
+
+    TORCH_INTERNAL_ASSERT(stack.size() == 1, "A boxed kernel should only push one return to the stack");
+    return std::move(stack[0]).to<Return>();
```

Comment:
Isn't `stack[0]` already a temporary, so the `std::move` is not necessary?

Reply:
no, `std::vector<T>::operator[]` returns a `T&`, not a `T`.